app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt

```diff
@@ -0,0 +1,376 @@
+package org.schabi.newpipe.util.potoken
+
+import android.content.Context
+import android.os.Build
+import android.os.Handler
+import android.os.Looper
+import android.util.Log
+import android.webkit.ConsoleMessage
+import android.webkit.JavascriptInterface
+import android.webkit.WebChromeClient
+import android.webkit.WebView
+import androidx.annotation.MainThread
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
+import io.reactivex.rxjava3.core.Single
+import io.reactivex.rxjava3.core.SingleEmitter
+import io.reactivex.rxjava3.disposables.CompositeDisposable
+import io.reactivex.rxjava3.schedulers.Schedulers
+import org.schabi.newpipe.BuildConfig
+import org.schabi.newpipe.DownloaderImpl
+import java.time.Instant
+
+class PoTokenWebView private constructor(
+    context: Context,
+    // to be used exactly once only during initialization!
+    private val generatorEmitter: SingleEmitter<PoTokenGenerator>,
+) : PoTokenGenerator {
+    private val webView = WebView(context)
+    private val disposables = CompositeDisposable() // used only during initialization
+    private val poTokenEmitters = mutableListOf<Pair<String, SingleEmitter<String>>>()
+    private lateinit var expirationInstant: Instant
+
+    //region Initialization
+    init {
+        val webviewSettings = webView.settings
+        //noinspection SetJavaScriptEnabled we want to use JavaScript!
+        webviewSettings.javaScriptEnabled = true
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            webviewSettings.safeBrowsingEnabled = false
+        }
```

Comment:
The [compat version](https://developer.android.com/reference/androidx/webkit/WebSettingsCompat#setSafeBrowsingEnabled(android.webkit.WebSettings,boolean)) of this method might be needed to support older versions of Android.

More information: https://www.youtube.com/watch?v=7Wq8Lwt_Cbg